datastore/redis.go

```diff
@@ -800,6 +811,65 @@ func (r *RedisCache) SetFloorBidValue(slot uint64, parentHash, proposerPubkey, v
 	return err
 }
 
+// BeginProcessingSlot signals that a builder process is handling blocks for a given slot
+func (r *RedisCache) BeginProcessingSlot(ctx context.Context, slot uint64) (err error) {
+	// Should never process more than one slot at a time
+	if r.currentSlot != 0 {
+		return fmt.Errorf("already processing slot %d", r.currentSlot) //nolint:goerr113
+	}
+
+	keyProcessingSlot := r.keyProcessingSlot(slot)
+
+	pipe := r.client.TxPipeline()
+	pipe.Incr(ctx, keyProcessingSlot)
+	pipe.Expire(ctx, keyProcessingSlot, expiryLock)
+	_, err = pipe.Exec(ctx)
+
+	if err != nil {
+		return err
+	}
+
+	r.currentSlot = slot
+	return nil
+}
+
+// EndProcessingSlot signals that a builder process is done handling blocks for the current slot
+func (r *RedisCache) EndProcessingSlot(ctx context.Context) (err error) {
+	// Do not decrement if called multiple times
+	if r.currentSlot == 0 {
+		return nil
+	}
+
+	keyProcessingSlot := r.keyProcessingSlot(r.currentSlot)
+
+	pipe := r.client.TxPipeline()
+	pipe.Decr(ctx, keyProcessingSlot)
+	pipe.Expire(ctx, keyProcessingSlot, expiryLock)
+	_, err = pipe.Exec(ctx)
+
+	if err != nil {
+		return err
+	}
+
+	r.currentSlot = 0
+	return nil
+}
+
+// WaitForSlotComplete waits for a slot to be completed by all builder processes
+func (r *RedisCache) WaitForSlotComplete(ctx context.Context, slot uint64) (err error) {
+	keyProcessingSlot := r.keyProcessingSlot(slot)
+	for {
```

Comment:
wondering if there should be a timeout if `EndProcessingSlot` fails to decrement the processing slot and the loop is stuck waiting for the slot to complete

Reply:
The current failsafe is that the key is slot-specific and has a fairly short expiry. The loop will break on a value of 0, key expiry, or any other redis error. But it probably wouldn't hurt to have an explicit timeout on the mev-boost-relay side. I'll whip something up.

Also can consider keyspace events (https://redis.io/docs/manual/keyspace-notifications/) as a cleaner way to accomplish this, but might require some redis server config.